src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter_1_1.kt

```diff
@@ -329,10 +357,39 @@ class IonRawBinaryWriter_1_1 internal constructor(
     private inline fun writeScalar(valueWriterExpression: () -> Int) = openValue {
         val numBytesWritten = valueWriterExpression()
         currentContainer.length += numBytesWritten
+        currentContainer.numChildren++
+    }
+
+    /**
+     * Helper function for writing scalar values that could be tagless.
+     *
+     * @param ifTagged should be a function that writes the scalar value to the buffer, and returns the number of bytes that were written.
+     * @param ifTagless should be a function that writes the scalar value to the buffer _without an opcode_ and returns the number of bytes that were written.
+     */
+    private inline fun writeTaggedOrTaglessScalar(
+        ifTagged: () -> Int,
+        ifTagless: (PrimitiveType) -> Int,
+    ) {
+        val primitiveType = when (currentContainer.type) {
+            EEXP -> presenceBitmapStack.peek().signature[currentContainer.numChildren].type.primitiveType
+            EXPR_GROUP -> currentContainer.primitiveType
+            else -> null
```

Comment:
Could this be:
```kt
else -> return writeScalar { taggedEncoder() }
```
to simplify what follows?

Reply:
No, because it's possible that `currentContainer.primitiveType` is `null`.